src/hotspot/share/opto/addnode.cpp

```diff
@@ -769,6 +771,597 @@ uint AddPNode::match_edge(uint idx) const {
+MergeLoadInfo MergePrimitiveLoads::merge_load_info(LoadNode* load) const {
+  const MergeLoadInfo invalid = MergeLoadInfo();
+  const Node* check = bypass_i2l(load);
+  Node* combine_oper = nullptr;
+  jint shift = -1;
+
+  // Check the Load node has the pattern "(Or (LShift (Load .. ) ConI) ..)" or "(Or (Load ..) ..)"
+  for (DUIterator_Fast imax, iter = check->fast_outs(imax); iter < imax; iter++) {
+    Node *out = check->fast_out(iter);
+    switch (out->Opcode()) {
+      case Op_OrI:
+      case Op_OrL: {
+        // match pattern: (Or (Load ..) ..)
+        if (combine_oper == nullptr) {
+          combine_oper = out;
+          shift = 0;
+        } else {
+          // Too much Or usages
+          return invalid;
+        }
+        break;
+      }
+      case Op_LShiftI:
+      case Op_LShiftL: {
+        // match pattern: (Or (LShift (Load ..) ConI) ..)
+        Node* shift_oper = out->isa_LShift();
+        if (shift_oper->outcnt() != 1 ||                                               // Shift should has only one usage
+            !is_supported_combine_opcode(shift_oper->unique_out()->Opcode()) ||    // Not used by combine operator
+            !shift_oper->in(2)->is_ConI()) {                                         // Not shift by constant
+          return invalid;
+        }
+        if (combine_oper == nullptr) {
+          combine_oper = shift_oper->unique_out();
+        } else {
+          // Too much combine operators
+          return invalid;
+        }
+        shift = shift_oper->in(2)->as_ConI()->get_int();
+        if (shift % (load->memory_size() * BitsPerByte) != 0) {
+          // Shift value is not aligned with memory size
+          return invalid;
+        }
+        break;
+      }
+      default:
+        // can not handle other usage
+        return invalid;
+    }
+  }
+  if (combine_oper == nullptr) {
+    return invalid;
+  }
+
+  assert(shift != -1, "must be set");
+  return MergeLoadInfo(load, combine_oper, shift);
+}
+
+Node* MergePrimitiveLoads::run() {
+  if (_combine->is_merge_memops_checked() || !has_no_merge_load_combine_below()) {
+    // no progress
+    return nullptr;
+  }
+  _combine->set_merge_memops_checked(true);
+
+  // go up through combine operators to find load node
+  LoadNode* load = nullptr;
+  Node* oper = _combine;
+  NOT_PRODUCT(int steps = 0;)    // prevent dead loop in bad graph
+  while (load == nullptr NOT_PRODUCT(&& steps < 30)) {
+    assert(is_supported_combine_opcode(oper->Opcode())
+           || oper->Opcode() == Op_ConvI2L || oper->is_LShift(), "unexpected node");
+    Node* lhs = oper->in(1); // Check one input is enough
+    assert(lhs != nullptr, "sanity");
+    if (lhs->Opcode() == Op_ConvI2L || lhs->is_LShift() || lhs->Opcode() == _combine->Opcode()) {
+      oper = lhs;
+      continue;
+    } else if (is_supported_load_opcode(lhs->Opcode())) {
+      load = lhs->as_Load();
+      break;
+    } else {
+      // not found
+      add_operators_to_worklist(_combine);
```

Comment:
Why are you doing this? If an input still needs to be transformed, then it should be put onto the work list by the inputs of that operator. And not by `combine`, i.e. the use of that operator.

Plus: if those operators are now transformed, would we actually ever get back here and attempt optimizing again? Your flag is now already set with `set_merge_memops_checked`, so we would not get here again, right?

Reply:
It's in the process of transforming `_combine`, so I think it follows the rule, "put inputs of the operator to worklist", do I misunderstand?

Yes, "_mege_memops_checked" flag is used to block re-transforming `_combine` again.